File: pkg/models/glsa.go
package models

import (
	"glsamaker/pkg/database/connection"
	"glsamaker/pkg/models/bugzilla"
	"glsamaker/pkg/models/cve"
	"glsamaker/pkg/models/gpackage"
	"glsamaker/pkg/models/users"
	"time"
)

type Glsa struct {
	//	Id                string
	Id          int64 `pg:",pk,unique"`
	Alias       string
	Type        string
	Title       string
	Synopsis    string
	Packages    []gpackage.Package
	Description string
	Impact      string
	Workaround  string
	Resolution  string
	References  []Reference
	Permission  string
	Access      string
	Severity    string
	Keyword     string
	Background  string
	Bugs        []bugzilla.Bug `pg:"many2many:glsa_to_bugs,joinFK:bug_id"`
	Comments    []cve.Comment  `pg:",fk:glsa_id"`
	Revision    string
	ApprovedBy  []int64
	DeclinedBy  []int64
	CreatorId   int64
	Creator     *users.User
	Created     time.Time
	Updated     time.Time
	Status      Status `pg:"-"`
}

type GlsaToBug struct {
	GlsaId int64 `pg:",unique:glsa_to_bug"`
	BugId  int64 `pg:",unique:glsa_to_bug"`
}

type Reference struct {
	Title string
	URL   string
}

type Status struct {
	BugReady       bool
	Approval       string
	WorkflowStatus string
	Permission     string
}

func (glsa *Glsa) IsBugReady() bool {
	bugReady := true
	for _, bug := range glsa.Bugs {
		bugReady = bugReady && bug.IsReady()
	}
	return bugReady
}

func (glsa *Glsa) ComputeStatus(user *users.User) {
	status := Status{
		BugReady:       glsa.IsBugReady(),
		Approval:       "none",
		WorkflowStatus: "todo",
		Permission:     glsa.Permission,
	}

	if glsa.DeclinedBy != nil && len(glsa.DeclinedBy) > 0 {
		status.Approval = "declined"
	} else if glsa.ApprovedBy != nil && len(glsa.ApprovedBy) > 0 {
		status.Approval = "approved"
	} else if glsa.Comments != nil && len(glsa.Comments) > 0 {
		status.Approval = "comments"
	}

	if glsa.CreatorId == user.Id {
		status.WorkflowStatus = "own"
	} else if contains(glsa.ApprovedBy, user.Id) {
		status.WorkflowStatus = "approved"
	} else {
		for _, comment := range glsa.Comments {
			if comment.User == user.Id {
				status.WorkflowStatus = "commented"
				break
			}
		}
	}

	glsa.Status = status
}

func (glsa *Glsa) ComputeCommentBadges() {
	for _, comment := range glsa.Comments {
		user := new(users.User)
		connection.DB.Model(user).Where("id = ?", comment.User).Select()

		comment.UserBadge = user.Badge
	}
}

func contains(arr []int64, element int64) bool {
	for _, a := range arr {
		if a == element {
			return true
		}
	}
	return false
}

File: src/models/glsa.py
from app import db
from models.bug import Bug
from models.reference import Reference
from models.user import User
from models.package import Package


glsa_to_bug = db.Table('glsa_to_bug',
                       db.Column('glsa_id', db.String(),
                                 db.ForeignKey('glsa.glsa_id')),
                       db.Column('bug_id', db.String(),
                                 db.ForeignKey('bug.bug_id')))

glsa_to_ref = db.Table('glsa_to_ref',
                       db.Column('glsa_id', db.String(),
                                 db.ForeignKey('glsa.glsa_id')),
                       db.Column('ref_text', db.String(),
                                 db.ForeignKey('reference.ref_text')))

glsa_to_affected = db.Table('glsa_to_affected',
                            db.Column('glsa_id', db.String(),
                                      db.ForeignKey('glsa.glsa_id')),
                            db.Column('affected_id', db.Integer(),
                                      db.ForeignKey('affected.affected_id')))



class GLSA(db.Model):
    __tablename__ = 'glsa'

    glsa_id = db.Column(db.String(), primary_key=True)
    title = db.Column(db.String())
    synopsis = db.Column(db.String())
    product_type = db.Column(db.Enum('ebuild', 'infrastructure',
                                     name='product_types'))
    product = db.Column(db.String())
    announced = db.Column(db.String())
    revision_count = db.Column(db.Integer())
    revised_date = db.Column(db.Date())
    bugs = db.relationship("Bug", secondary=glsa_to_bug)
    # TODO: lots of variation here, we should cleanup eventually
    access = db.Column(db.Enum('unknown', 'local', 'remote',
                               'local, remote',
                               'remote, local',
                               'local and remote',
                               # glsa-200503-10
                               'remote and local',
                               # glsa-200311-0{1,2}
                               'local / remote',
                               # glsa-200403-01
                               'local and remote combination',
                               # glsa-201001-03, glsa-201011-01
                               'local remote',
                               # glsa-200408-18
                               'remote root',
                               # glsa-200401-04
                               'man-in-the-middle',
                               # glsa-200312-01
                               '',
                               name='access_types'))
    affected = db.relationship("Affected", secondary=glsa_to_affected)
    background = db.Column(db.String())
    description = db.Column(db.String())
    impact_type = db.Column(db.Enum('minimal', 'low', 'medium',
                                    'normal', 'high',
                                    name='impact_types'))
    impact = db.Column(db.String())
    workaround = db.Column(db.String())
    resolution = db.Column(db.String())
    references = db.relationship("Reference", secondary=glsa_to_ref)
    # TODO: bugReady metadata tag?
    requester = db.Column(db.Integer, db.ForeignKey(User.id))
    submitter = db.Column(db.Integer, db.ForeignKey(User.id))
    requested_time = db.Column(db.DateTime())
    submitted_time = db.Column(db.DateTime())

    def get_bugs(self):
        return [bug.bug_id for bug in self.bugs]

    def get_references(self):
        return [ref.ref_text for ref in self.references]
